lib/promscrape/discovery/kubernetes/endpointslices.go

```diff
@@ -0,0 +1,214 @@
+	if svc != nil {
+		svc.appendCommonLabels(m)
+	}
+	if ea.TargetRef.Kind != "Pod" || p == nil {
+		return m
+	}
+	p.appendCommonLabels(m)
+	for _, c := range p.Spec.Containers {
+		for _, cp := range c.Ports {
+			if cp.ContainerPort == epp.Port {
+				p.appendContainerLabels(m, c, &cp)
+				podPortsSeen[p] = append(podPortsSeen[p], cp.ContainerPort)
+				break
+			}
+		}
+	}
+
+	return m
+}
+
+// //getEndpointSliceLabels builds labels for given EndpointSlice
+func getEndpointSliceLabels(eps *EndpointSlice, addr string, ea Endpoint, epp EndpointPort) map[string]string {
+
+	var status bool
+	if ea.Conditions.Ready != nil && *ea.Conditions.Ready {
+		status = true
+	}
+	addr = discoveryutils.JoinHostPort(addr, epp.Port)
+	m := map[string]string{
+		"__address__":                                               addr,
+		"__meta_kubernetes_namespace":                               eps.Metadata.Namespace,
+		"__meta_kubernetes_endpointslice_name":                      eps.Metadata.Name,
+		"__meta_kubernetes_endpointslice_address_type":              eps.AddressType,
+		"__meta_kubernetes_endpointslice_endpoint_conditions_ready": strconv.FormatBool(status),
+		"__meta_kubernetes_endpointslice_port_name":                 epp.Name,
+		"__meta_kubernetes_endpointslice_port_protocol":             epp.Protocol,
+		"__meta_kubernetes_endpointslice_port_app_protocol":         epp.AppProtocol,
+		"__meta_kubernetes_endpointslice_port":                      strconv.FormatUint(uint64(epp.Port), 10),
+	}
+	if ea.TargetRef.Kind != "" {
+		m["__meta_kubernetes_endpointslice_address_target_kind"] = ea.TargetRef.Kind
+		m["__meta_kubernetes_endpointslice_address_target_name"] = ea.TargetRef.Name
+	}
+	if ea.Hostname != "" {
+		m["__meta_kubernetes_endpointslice_endpointslice_endpoint_hostname"] = ea.Hostname
+	}
+	for k, v := range ea.Topology {
+		m["__meta_kubernetes_endpointslice_endpoint_topology_"+discoveryutils.SanitizeLabelName(k)] = v
+		m["__meta_kubernetes_endpointslice_endpoint_topology_present_"+discoveryutils.SanitizeLabelName(k)] = "true"
+	}
+	return m
+}
+
+// EndpointSliceList - implements kubernetes endpoint slice list object,
+// that groups service endpoints slices.
+// https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#endpointslice-v1beta1-discovery-k8s-io
+type EndpointSliceList struct {
+	Items []EndpointSlice
+}
+
+// EndpointSlice - implements kubernetes endpoint slice.
+// https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#endpointslice-v1beta1-discovery-k8s-io
+type EndpointSlice struct {
+	Metadata    ObjectMeta
+	Endpoints   []Endpoint
+	AddressType string
+	Ports       []EndpointPort
+}
+
+// Endpoint implements kubernetes object endpoint for endpoint slice.
+// https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#endpoint-v1beta1-discovery-k8s-io
+type Endpoint struct {
+	Addresses  []string
+	Conditions *EndpointConditions
```

Comment:
Probably it would be better storing `Conditions` field by value instead of by pointer. Then `nil` check on line 155 won't be needed.

Reply:
done